(Bugfix) Make sure one to one entities are encoded correctly by the JsonApiEncoder

AlexBachmann opened the pull request:
### 1. Why is this change necessary?
If one of our entities (Entity A) has a one to one association with a different entity (Entity B), it is very likely that
this entity has a relationship back to the first entity.

Entity A  < ------ > Entity B

Searching for Entity A, while also loading Entity B ($criteria->addAssociation('entityB') would result in a result like this

$entiyA
|-- $entiyB
|-- |-- $entityA
|-- |-- |-- $entityB = null

Or, if $entityB is implemented via an entity extension:

$entityA
|-- $extensions
|-- |-- $entityB
|-- |-- |-- $entityA
|-- |-- |-- |-- $extensions
|-- |-- |-- |-- |-- $entiyB = null

The problem we have in these situations is the following, entity A is serialized first in the process of serializing entityB:

Process:
Serialize Entity A --> Serialize Relationships / Extensions of Entity A --> Serialize Entity B (which is related / an extension of Entity A) --> Serialize relationships / Extensions of Entity B --> Serialize Entity A (which is a related / an extension of Entity B)

This process naturally ends at some point, as we are walking down the tree. HOWEVER, we do have a problem once we add the serialized data to the included section of the record.
Since we include All relationships / extensions of Entity A the first time in the process of serializing Entity B, we do not add the data again when we come back to Entity A when we walk back up the tree. This results in the fact that the data value of Entity B remains NULL in the included part, even though the association has been loaded in the search query.

This phenomenon is easier understood running the unit test, which i provided with this commit. Without the fix, the toOne Relationship is not correctly serialized.

### 2. What does this change do, exactly?

This change merges entity data of entities that have been previously added instead of ignoring the entity. By merging the entity data, we make sure that those entities that have been hydrated the most are included in the response.

### 3. Describe each step to reproduce the issue or behaviour.

See #1


### 4. Please link to the relevant issues (if any).


### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [ ] I have created a [changelog file](https://github.com/shopware/platform/blob/master/adr/2020-08-03-Implement-New-Changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

JoshuaBehrens commented:
Dude, that situation. Good explanation

shyim commented on src/Core/Framework/Test/Api/Serializer/JsonApiEncoderTest.php:
```diff
@@ -26,15 +31,73 @@
+            DROP TABLE IF EXISTS `extended_product`;
+            CREATE TABLE `extended_product` (
+                `id` BINARY(16) NOT NULL,
+                `name` VARCHAR(255) NULL,
```
Can you add the charset?

AlexBachmann replied:
done

bneumann97 commented:
Hey @AlexBachmann, thanks for your contribution and the test 💙  Could you please add a [changelog file](https://github.com/shopware/platform/blob/master/adr/2020-08-03-Implement-New-Changelog.md)? 🙂

bneumann97 commented:
After some internal discussions we have some question regarding your PR.
- Did you make sure that this is not an infinite loop?
- What about the encoding performance?
- And also we would need more test cases than the one you provided for example `ToMany` associations.

AlexBachmann commented:
Hey Bastian,

thank you for your  response. But I want to point out something here (read this in a friendly speaking voice): This is a bug, not a feature request. It should be in your interest, to fully qualify the scope of the bug and not hand it down to a third party contributor. I mean, I will answer your  questions, but what does it matter what I say? You need to verify these questions yourselves. This is your JSON API Encoder, one of the most essential parts of your application.

Concerning your questions:
**Did you make sure, that this is not an infinite loop?** When you look at the solution I suggested, you see there is no way, this could end in an infinite loop. The infinite loop _could_ occur in the JsonApiEncoder itself (which I have not changed), but since it only walks up and down the tree, this does not happen

**What about the encoding performance?** This is a valid question, which I have thought about. Merging the entities DOES cause some overhead. However, the  caching strategy (only use the first assignment of the included entity) does not work, because, as I demonstrated with the unit test, the first time the entity is included, it is not fully hydrated. An alternative strategy (instead of merging the entities) could be to always include the LAST entity assignment. This would reduce the overhead of merging the entities, however it needs to be evaluated if this does not cause other problems down the road (what happens if we hydrate the entity in an association, and not in the root entity?). I figured, that merging the entities always results in including the entity record that is the most hydrated, which is desirable. I find the overhead acceptable, because the merging ONLY takes place, if the same entity has been included before (which happens really rarely).

**And also we would need more test cases than the one you provided for example ToMany associations** Now that I think about it, the problem might also occur with `toMany` associations (needs to be checked). The provided fix should work independently from the type of association, since it only changes the way, entites are added to the included part of the record.  I'll try add some tests if I find the time. However, I have to stress this again: This is a bug, which really should be looked at and understood by you.

bneumann97 commented:
Hey @AlexBachmann, I just checked out your branch and compared it with our trunk. Your change looks good also performance wise. It is slower but only a very little. If you could add some more test cases, your PR would be ready to go! 🙂 💙

AlexBachmann commented:
Added a new test for entity extensions with toMany associations. Added a changelog entry.

bneumann97 commented on src/Core/Framework/Test/Api/Serializer/JsonApiEncoderTest.php:
```diff
@@ -140,6 +203,101 @@ public function testEncodeStructWithToManyExtension(): void
+        /** @var ProductEntity $product */
+        $product = $this->productRepository->search($criteria, Context::createDefaultContext())->get($productId);
+        $encoder = $this->getContainer()->get(JsonApiEncoder::class);
+        $encodedResponse = $encoder->encode(new Criteria(), $productDefinition, $product, SerializationFixture::API_BASE_URL, SerializationFixture::API_VERSION);
```
@AlexBachmann your tests seem to fail. The `encode()` method expects an array of metadata as 5th parameter, but you give an int. Can you have a look at that? https://github.com/shopware/platform/blob/832f7282064cb47c7e32a2d60a8d3f2bff9030a8/src/Core/Framework/Api/Serializer/JsonApiEncoder.php#L35

AlexBachmann replied:
That's right. The signature of the method changed since i started the pull request. I updated the tests so they should now pass.

AlexBachmann committed:
Make sure one to one entities are encoded correctly by the JsonApiEncoder

If one of our entities (Entity A) has a one to one association with a different entity (Entity B), it is very likely that
this entity has a relationship back to the first entity.

Entity A < ------ > Entity B

Searching for Entity A, while also loading Entity B ($criteria->addAssociation('entityB') would result in a result like this

$entiyA
|-- $entiyB
|-- |-- $entityA
|-- |-- |-- $entityB = null

Or, if $entityB is implemented via an entity extension:

$entityA
|-- $extensions
|-- |-- $entityB
|-- |-- |-- $entityA
|-- |-- |-- |-- $extensions
|-- |-- |-- |-- |-- $entiyB = null

The problem we have in these situations is the following, entity A is serialized first in the process of serializing entityB:

Process:
Serialize Entity A --> Serialize Relationships / Extensions of Entity A --> Serialize Entity B (which is related / an extension of Entity A) --> Serialize relationships / Extensions of Entity B --> Serialize Entity A (which is a related / an extension of Entity B)

This process naturally ends at some point, as we are walking down the tree. HOWEVER, we do have a problem once we add the serialized data to the included section of the record.
Since we include All relationships / extensions of Entity A the first time in the process of serializing Entity B, we do not add the data again when we come back to Entity A when we walk back up the tree. This results in the fact that the data value of Entity B remains NULL in the included part, even though the association has been loaded in the search query.

This phenomenon is easier understood running the unit test, which i provided with this commit. Without the fix, the toOne Relationship is not correctly serialized.

bneumann97 commented:
Hey @AlexBachmann, thanks for your contribution 💙
